geode-core/src/main/java/org/apache/geode/redis/internal/GeoCoder.java

```diff
@@ -0,0 +1,517 @@
+        coord.add(Double.toString(element.getCoord().get().getLatitude()));
+      }
+
+      String hash = "";
+      if (element.getHash().isPresent()) {
+        hash = element.getHash().get();
+      }
+
+      if (distStr != "" || !coord.isEmpty() || hash != "") {
+        List<Object> elementData = new ArrayList<>();
+        elementData.add(name);
+        if (distStr != "")
+          elementData.add(distStr);
+        if (!coord.isEmpty())
+          elementData.add(coord);
+        if (hash != "")
+          elementData.add(hash);
+
+        responseElements.add(elementData);
+      } else {
+        responseElements.add(name);
+      }
+    }
+
+    return Coder.getBulkStringArrayResponse(alloc, responseElements);
+  }
+
+  /**
+   * Converts geohash to lat/long.
+   *
+   * @param hash geohash as base32
+   * @return a GeoCoord object containing the coordinates
+   */
+  public static GeoCoord geoPos(char[] hash) {
+    Pair<char[], char[]> hashBits = deinterleave(hash);
+
+    return new GeoCoord(coord(hashBits.fst, LONG_MIN, LONG_MAX),
+        coord(hashBits.snd, LAT_MIN, LAT_MAX));
+  }
+
+  /**
+   * Calculates distance between two points.
+   *
+   * @param hash1 geohash of first point
+   * @param hash2 geohash of second point
+   * @return distance in meters
+   */
+  public static Double geoDist(char[] hash1, char[] hash2) {
+    GeoCoord coord1 = geoPos(hash1);
+    GeoCoord coord2 = geoPos(hash2);
+
+    Double lat1 = Math.toRadians(coord1.getLatitude());
+    Double long1 = Math.toRadians(coord1.getLongitude());
+    Double lat2 = Math.toRadians(coord2.getLatitude());
+    Double long2 = Math.toRadians(coord2.getLongitude());
+
+    return dist(long1, lat1, long2, lat2);
+  }
+
+  /**
+   * Calculates geohash given latitude and longitude as byte arrays encoding decimals.
+   *
+   * @param lon byte array encoding longitude as decimal
+   * @param lat byte array encoding latitude as decimal
+   * @return geohash as base32
+   */
+  public static String geoHash(byte[] lon, byte[] lat, int length) throws CoderException {
+    return bitsToHash(geoHashBits(lon, lat, length));
+  }
+
+  public static char[] geoHashBits(byte[] lon, byte[] lat, int length) throws CoderException {
+    Double longitude = Coder.bytesToDouble(lon);
+    Double latitude = Coder.bytesToDouble(lat);
+
+    int longLen, latLen;
+    if (length % 2 == 0) {
+      longLen = length / 2;
+      latLen = length / 2;
+    } else {
+      longLen = (length + 1) / 2;
+      latLen = (length - 1) / 2;
+    }
+
+    char[] longDigits = coordDigits(longitude, LONG_MIN, LONG_MAX, longLen);
+    char[] latDigits = coordDigits(latitude, LAT_MIN, LAT_MAX, latLen);
+
+    return interleave(longDigits, latDigits);
+  }
+
+  /**
+   * Return a set of hashes (center + 8) that are able to cover a range query
+   * for the specified position and radius.
+   */
+  public static HashNeighbors geoHashGetAreasByRadius(double longitude, double latitude,
```

Comment:
This looks pretty similar to the corresponding code in Redis including the comments. There might be some issues here around taking this without including the original license text. I'd suggest that we find a clear answer to this before we merge this.